### PR TITLE
feature: Publish E2E tests to npmjs for dev builds

### PR DESCRIPTION
**Pull request by mkuznyetsov**
Signed-off-by: Mykhailo Kuznietsov <mkuznets@redhat.com>

### What does this PR do?
- move location on E2E dockerfiles to the tests/e2e directory
- Add the job that will perform publishing of E2E npm package to npmjs. 
  - for each commit to main branch, the tag will be "next" and the version will be suffixed with `dev-<git_commit_hash>`  - "7.41.0-dev-abcd123"
  - for each commit to bugfix branch (e.g. 7.40.x), the tag will be of the bugfix branch name "7.40.x", and the version will be suffixed with `dev-<git_commit_hash>`  - "7.41.3-dev-abcd123"
 - commit package-lock.json, as it will be required for the use of `npm ci` command 
 
  

### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/18773
https://github.com/eclipse/che/issues/18772

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix or new feature ](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix-or-new-feature)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: .github/workflows/typescript-publish.yml [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        if [[ ${GITHUB_REF##*/} == "7."**".x" ]]; then
+          echo "[INFO] using ${GITHUB_REF##*/} tag"
+          DIST_TAG=${GITHUB_REF##*/}
+        else
+          echo "[INFO] using "next" tag"
+          DIST_TAG=next
+        fi
+
+        cd tests/e2e
+        npm ci && npm run tsc
```
Comment (ScrewTSW) [inline .github/workflows/typescript-publish.yml:48]: So, will we be doing the `npm audit --fix` and pushing package-lock.json manually?

Comment (mkuznyetsov) [reply .github/workflows/typescript-publish.yml:48]: I'd say yes, we can do it once, and then package-lock.json will stay untouched, until someone else will have to modify it (by using npm install command), or am I wrong?

Comment (rhopp) [reply .github/workflows/typescript-publish.yml:48]: I would vote for package-lock.json being updated "manually" by test developers and anything in CI would use `npm  ci`.
The "natural" update of package-lock.json will happen by merging devs PRs (Devs will naturally do "npm install" locally, which should update package-lock.json and this one should be commited&pushed together with their PR).

As for `npm audit` - From what I read `npm audit fix` will update just within the ranges defined in package.json. To properly udpate deps to resolve the vulnerabilities we would need to run `npm audit fix --force` which updates package.json as well.

File: tests/e2e/package-lock.json [hunk 2]
```diff
@@ -1,13 +1,2532 @@
...
-  "version": "1.0.0",
-  "lockfileVersion": 1,
+  "name": "@eclipse-che/che-e2e",
+  "version": "7.39.0",
+  "lockfileVersion": 2,
   "requires": true,
+  "packages": {
+    "": {
+      "name": "@eclipse-che/che-e2e",
+      "version": "7.39.0",
```
Comment (dmytro-ndp) [inline tests/e2e/package-lock.json:9]: shouldn't it be 7.40.0 ?

**Comment by mkuznyetsov** [comment]
@dmytro-ndp well there's gotta be some version in it, so might as well put the last released version in it (I did so as https://github.com/eclipse/che/issues/18771 required)
And 7.40.0 values will be updated with the release.yml workflow (after Che 7.40.0 E2E will be released)

**Comment by mkuznyetsov** [comment]
@dmytro-ndp Instead of dummy version, we can stick with *-SNAPSHOT* suffix, like it's done in several other npm projects, such as our registries. That way, it would display what's the potential next version gonna be

File: tests/e2e/build/dockerfiles/README.md [hunk 3]
```diff
@@ -14,10 +14,10 @@ installed Docker on your machine.
 These tests serves for testing Che 7 happy path. You can see description about these tests here: https://github.com/eclipse/che/tree/master/tests/e2e.
 
 ## How to run it
-The easiest way is to run them via Docker. To build locally you have go to ` dockerfiles ` folder and execute following command: 
+The easiest way is to run them via Docker. To build locally you have execute the following command:
```
Comment (Balajitestingworld) [inline tests/e2e/build/dockerfiles/README.md:17]:
```suggestion
The easiest way is to run them via Docker. To build locally you have to execute the following command: 
```

Comment (makambalaji) [inline tests/e2e/build/dockerfiles/README.md]:
```suggestion
The easiest way is to run them via Docker. To build locally you have to execute the following command: 
```

File: .github/workflows/typescript-publish.yml (hunk 1 above)
Comment (makambalaji) [reply .github/workflows/typescript-publish.yml:48]: @mkuznyetsov - Please execute `npm audit fix --force` and update the package.json, package.lock.json. So we are good to merge the pr.
cc: @dmytro-ndp

Comment (makambalaji) [reply .github/workflows/typescript-publish.yml:48]: @mkuznets - On `npm audit fix --force` the mocha version is updating from `7.1.1` to `9.1.3`,  Where opts files are [deprecated](https://main-jenkins-csb-crwqe.apps.ocp4.prod.psi.redhat.com/blue/organizations/jenkins/Che%2Fe2e%2Fminikube%2Fpr-check%2Fche-tests-pr-check/detail/che-tests-pr-check/287/pipeline/). IMO its better to have a separate pr for these changes.
cc: @dmytro-ndp , @rhopp

**Comment by makambalaji** [comment]
LGTM

**Comment by makambalaji** [comment]
LGTM

**Review by makambalaji — APPROVED** [review]

**Review by dmytro-ndp — APPROVED** [review]
Validation results:
- build of che-e2e was successful: quay.io/crw_pr/che-e2e:20864
- execution of Happy path E2E tests from host machine (`npm install`) - passed: https://main-jenkins-csb-crwqe.apps.ocp4.prod.psi.redhat.com/job/Che/job/e2e/job/minikube/job/basic/job/che-server/1231/
- execution of Happy path E2E tests from withing che-e2e image (`npm run`) - passed: https://main-jenkins-csb-crwqe.apps.ocp4.prod.psi.redhat.com/job/Che/job/e2e/job/minikube/job/basic/job/che-server/1232/

**Pull request merged**